### PR TITLE
Support `GatherNd` op using two-layers `Loop` and `Gather`

### PR DESCRIPTION
**Pull request by lucienwang1009**
- Support `GatherNd` op using two-layers `Loop` and `Gather`
- Disable `GatherNd` tests unless with the target "rs6"

### CODE REVIEW

File: tf2onnx/function/gathernd.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                   helper.make_tensor_value_info("params", node_dtype, [])]
+    body_outputs = [helper.make_tensor_value_info("cond_out", onnx_pb.TensorProto.BOOL, [],),
+                    helper.make_tensor_value_info("params_out", node_dtype, []),
+                    helper.make_tensor_value_info("result", node_dtype, [])]
+    body_nodes = []
+    index = ctx.make_node("Gather", [flatten_indices.output[0], "i"], attr={"axis": 0})
+    index_squeeze = ctx.make_node("Squeeze", [index.output[0]], attr={"axes": [0]})
+    # inner loop to gather result
+    inner_loop_nodes, inner_loop = make_gathernd_inner_loop(ctx, "params", index_squeeze, node_dtype)
+    body_nodes.extend([index.op, index_squeeze.op])
```
Comment (pengwa) [inline tf2onnx/function/gathernd.py]: combine into an extend?

File: tf2onnx/function/gathernd.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                                  [outter_shape_sum.output[0], inner_shape.output[0]],
+                                  attr={"axis": 0},
+                                  dtypes=[TensorProto.INT64])
+    flatten_indices = ctx.make_node("Reshape", [indices, flatten_shape.output[0]])
+    nodes.extend([indices_shape, outter_shape, inner_shape, outter_shape_sum, flatten_shape, flatten_indices])
+    # outter loop for each index
+    # for (int i=0; i<outter_shape_sum; i++) inner_loop(params, flatten_indices[i])
+    cond_const = ctx.make_const(utils.make_name("cond"), np.ones((), dtype=np.bool))
+    dummy_const = ctx.make_const(utils.make_name("dummy"), np.ones((), dtype=np.int64))
+    body_inputs = [helper.make_tensor_value_info("i", onnx_pb.TensorProto.INT64, []),
```
Comment (pengwa) [inline tf2onnx/function/gathernd.py]: better not hardcode the input name, use make_name for "i", in case of naming conflict.

Comment (pengwa) [reply tf2onnx/function/gathernd.py]: same for output

File: tf2onnx/function/gathernd.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+# pylint: disable=useless-return,broad-except,logging-not-lazy,unused-argument,missing-docstring
+def make_gathernd_inner_loop(ctx, params, index, dtype):
+    # gather_cur = params
+    # for (int i=0; i<size(index); i++)
+    #   gather_res = gather(gather_cur, index[i])
+    nodes = []
+    trip_node = ctx.make_node("Size", [index.output[0]])
+    nodes.append(trip_node.op)
+    cond_const = ctx.make_const(utils.make_name("cond"), np.ones((), dtype=np.bool))
+    body_inputs = [helper.make_tensor_value_info("inner_i", onnx_pb.TensorProto.INT64, []),
```
Comment (pengwa) [inline tf2onnx/function/gathernd.py]: make_name for "inner_i"

File: tests/test_backend.py [hunk 4]
```diff
@@ -499,6 +504,24 @@ def test_gather(self):
         _ = tf.identity(x_, name=_TFOUTPUT)
         self._run_test_case([_OUTPUT], {_INPUT: x_val})
 
+    @unittest.skipIf(*support_op_with_target('rs6', 'GatherNd'))
+    def test_gathernd(self):
+        x_val = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=np.float32)
+        indices = np.array([[[0, 1], [1, 1]], [[1, 2], [0, 2]]], dtype=np.int32)
+        x = tf.placeholder(tf.float32, x_val.shape, name=_TFINPUT)
+        x_ = tf.gather_nd(x, tf.constant(indices))
```
Comment (nbcsm) [inline tests/test_backend.py]: suggest to add a few more cases that can cover some corner case of your implementation.

File: tf2onnx/function/gathernd.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    nodes = []
+    trip_node = ctx.make_node("Size", [index.output[0]])
+    nodes.append(trip_node.op)
+    cond_const = ctx.make_const(utils.make_name("cond"), np.ones((), dtype=np.bool))
+    trip_name = utils.make_name("i")
+    cond_name = utils.make_name("cond")
+    cond_out_name = utils.make_name("cond_out")
+    cur_name = utils.make_name("gather_cur")
+    result_name = utils.make_name("res")
+    body_inputs = [helper.make_tensor_value_info(trip_name, onnx_pb.TensorProto.INT64, []),
```
Comment (pengwa) [inline tf2onnx/function/gathernd.py]: use utils.make_onnx_inputs_outputs() please

File: tf2onnx/function/gathernd.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    cond_const = ctx.make_const(utils.make_name("cond"), np.ones((), dtype=np.bool))
+    dummy_const = ctx.make_const(utils.make_name("dummy"), np.ones((), dtype=np.int64))
+    trip_name = utils.make_name("i")
+    cond_name = utils.make_name("cond")
+    cond_out_name = utils.make_name("cond_out")
+    dummy_name = utils.make_name("dummy")
+    dummy_out_name = utils.make_name("dummy_out")
+    result_name = utils.make_name("res")
+    body_inputs = [helper.make_tensor_value_info(trip_name, onnx_pb.TensorProto.INT64, []),
+                   helper.make_tensor_value_info(cond_name, onnx_pb.TensorProto.BOOL, []),
```
Comment (pengwa) [inline tf2onnx/function/gathernd.py]: same here

File: tf2onnx/function/gathernd.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+                    helper.make_tensor_value_info(result_name, node_dtype, [])]
+    body_nodes = []
+    index = ctx.make_node("Gather", [flatten_indices.output[0], trip_name], attr={"axis": 0})
+    index_squeeze = ctx.make_node("Squeeze", [index.output[0]], attr={"axes": [0]})
+    # inner loop to gather result
+    inner_loop_nodes, inner_loop = make_gathernd_inner_loop(ctx, params, index_squeeze, node_dtype)
+    body_nodes.extend([index.op, index_squeeze.op] + inner_loop_nodes)
+    body_nodes.append(utils.make_onnx_identity(cond_name, cond_out_name))
+    body_nodes.append(utils.make_onnx_identity(dummy_name, dummy_out_name))
+    body_nodes.append(utils.make_onnx_identity(inner_loop.output[0], result_name))
```
Comment (pengwa) [inline tf2onnx/function/gathernd.py]: so frequent similar call. shall we do this way:

body_nodes.extend([index.op, index_squeeze.op] + inner_loop_nodes,
                                 utils.make_onnx_identity(cond_name, cond_out_name),
                                  utils.make_onnx_identity(dummy_name, dummy_out_name),
                                  ...)

File: tests/test_backend.py [hunk 8]
```diff
@@ -499,6 +504,41 @@ def test_gather(self):
...
+
+        x_val = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=np.float32)
+        indices = np.array([[0], [2], [4], [7]], dtype=np.int32)
+        x = tf.placeholder(tf.float32, x_val.shape, name=_TFINPUT)
+        x_ = tf.gather_nd(x, tf.constant(indices))
+        _ = tf.identity(x_, name=_TFOUTPUT)
+        self._run_test_case([_OUTPUT], {_INPUT: x_val})
+
+    @unittest.skipIf(*support_op_with_target('rs6', 'GatherNd'))
+    def test_gathernd_less_index(self):
```
[the remark this answers is not in the document]
Comment (lucienwang1009) [reply tests/test_backend.py]: it means index shape is less than params rank, so it will gather a vector rather than a scalar. for instance, in the first test, params (x_val) rank is 2 while index shape ([0]) is 1, it will gather [1,2,3] resemble with Gather.

File: tf2onnx/function/gathernd.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+tf2onnx.tf2onnx - gathernd op conversion
+"""
+import sys
+import numpy as np
+from onnx import helper, onnx_pb
+from onnx.onnx_pb import TensorProto
+from tf2onnx import utils
+
+
+# pylint: disable=useless-return,broad-except,logging-not-lazy,unused-argument,missing-docstring
```
[the remark this answers is not in the document]
Comment (lucienwang1009) [reply tf2onnx/function/gathernd.py]: some codes need them.

File: tf2onnx/function/gathernd.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    # reshape indices into [sum(indices[:-1]), indices[-1]]
+    indices_shape = ctx.make_node("Shape", [indices], dtypes=[TensorProto.INT64])
+    outter_shape = ctx.make_node("Slice",
+                                 [indices_shape.output[0]],
+                                 attr={"axes": [0], "ends": [-1], "starts": [0]},
+                                 dtypes=[TensorProto.INT64])
+    inner_shape = ctx.make_node("Slice",
+                                [indices_shape.output[0]],
+                                attr={"axes": [0], "ends": [sys.maxsize], "starts": [-1]},
+                                dtypes=[TensorProto.INT64])
```
[the remark this answers is not in the document]
Comment (lucienwang1009) [reply tf2onnx/function/gathernd.py]: good point! thanks!
